### DISCUSSION TITLE
Ordering of Scope(s) and Reduce in ReducerProtocol body

### DISCUSSION THREAD

**Discussion by mluisbrown** [pointfreeco/swift-composable-architecture#1761] [Q&A] [answered]
Just starting on migration our project to `ReducerProtocol` and have a question about ordering. Previously child reducers were combined through `pullback`s and necessarily came before the parent reducer, and this made changes in the child state visible to the parent reducer. Is this still the case with `RedcuerProtocol`? ie:

Should we do this:
```swift
var body: some ReducerProtocol {
  Scope(state: \.child1, action: /Action.child1) {
    Child1()
  }
  Scope(state: \.child2, action: /Action.child2) {
    Child2()
  }
  Reduce { state, action in 
    // parent reducer switch 
  }
}
```

or this:
```swift
var body: some ReducerProtocol {
  Reduce { state, action in 
    // parent reducer switch 
  }
  Scope(state: \.child1, action: /Action.child1) {
    Child1()
  }
  Scope(state: \.child2, action: /Action.child2) {
    Child2()
  }
}
```

The documentation has examples of both, although the documentation for `Scope` seems to imply that `Scope`s should come before the parent reducer.

Related to this question, we use the `onChange` reducer from IsoWords quite a lot and I was very pleased to see that there is version for `ReducerProtocol` in the IsoWords repo (https://github.com/pointfreeco/isowords/blob/main/Sources/TcaHelpers/OnChange.swift) which we are using. 

My question is, with the following code:

```swift
var body: some ReducerProtocol {
  Scope(state: \.child1, action: /Action.child1) {
    Child1()
  }
  Scope(state: \.child2, action: /Action.child2) {
    Child2()
  }
  Reduce { state, action in 
    // parent reducer switch 
  }
  .onChange(of: \.child1.foo) { ... }
}
```

will the `onChange` reducer detect changes made only by the parent `Reduce` or also changes made with the `Scope`d child reducers? It *seems* to be only detecting changes made in the parent `Reduce` but I haven't definitely been able to confirm that yet.

**Comment by mluisbrown** [comment, marked as answer]
> It seems to be only detecting changes made in the parent `Reduce` but I haven't definitely been able to confirm that yet.

I have now tested it and can confirm that `onChange` is only detecting changes made in the `Reduce` to which it is attached, not any changes made in the `Child1()` reducer.

**Reply by Alex293** [reply]
You can wrap your 3 reducers into one and attach onChange on the composed one.

**Reply by Alex293** [reply]
Yours reducer are run in the order you declare. One thing to be wary of is optional reducers but you can use .ifLet to be sure they are run first

**Reply by mluisbrown** [reply]
> You can wrap your 3 reducers into one and attach onChange on the composed one.

Ok, this works:

```swift
@ReducerBuilder<State, Action>
private func coreReducers() -> some ReducerProtocol<State, Action> {
  Scope(state: \.child1, action: /Action.child1) {
    Child1()
  }
  Scope(state: \.child2, action: /Action.child2) {
    Child2()
  }
  Reduce { state, action in 
    // parent reducer switch 
  }
}

var body: some ReducerProtocol {
  coreReducers()
    .onChange(of: \.child1.foo) { ... }
}
```

Thanks 👍 

You didn't reply in an "Answer" so I can't mark your answer as the correct one 😕

**Reply by tgrapperon** [reply]
As an aside, and there's nothing wrong with the computed properties, the dedicated tool for this operation is [`CombineReducers`](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Sources/ComposableArchitecture/Reducer/Reducers/CombineReducers.swift):
```swift
var body: some ReducerProtocolOf<Self> {
  CombineReducers {
    Scope(state: \.child1, action: /Action.child1) {
      Child1()
    }
    Scope(state: \.child2, action: /Action.child2) {
      Child2()
    }
    Reduce { state, action in 
      // parent reducer switch 
    }
  }
  .onChange(of: \.child1.foo) { ... }
}
```

**Reply by mluisbrown** [reply]
Ah, that's much nicer ❤️

**Reply by Alex293** [reply]
No worries mark yourself as the answer ! I would have suggested 
```
@ReducerBuilder<State, Action>
var coreReducers: some ReducerProtocol {
}
```
but indeed there is a nicer tool in the lib for that

**Reply by mluisbrown** [reply]
Plot twist: if I use `CombineReducers` the compiler gives up with `The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions`. Using a `@ReducerBuilder` computed property doesn't have this issue. FWIW this is a huge reducer 😄

**Reply by tgrapperon** [reply]
@mluisbrown There was an issue with this reducer on Xcode 14 beta, and a trick was used to help the compiler. Are you using the latest version of Xcode?

**Reply by mluisbrown** [reply]
@tgrapperon we're using Xcode 14.1, soon switching to 14.2.

**Reply by tgrapperon** [reply]
@mluisbrown OK, thank! I'll try to reproduce then.
